_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,565 @@
+      --producer-property 'sasl.login.callback.handler.class=io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler' 
+
+If you get a warning about credentials expiring really soon, you probably forgot to configure a longer access token lifespan, as instructed in [Setting up Keycloak](#setting-up-keycloak) section.
+
+One question we haven't answered yet is if authentication is in fact mandatory or if can we connect by using only TLS - without the client actually authenticating.
+
+Let's try with the consumer this time.
+
+    export PASSWORD=truststorepassword
+    export KAFKA_OPTS=" \
+      -Djavax.net.ssl.trustStore=/opt/kafka/certificates/kafka-client-truststore.p12 \
+      -Djavax.net.ssl.trustStorePassword=$PASSWORD \
+      -Djavax.net.ssl.trustStoreType=PKCS12"
+      
+    bin/kafka-console-consumer.sh --bootstrap-server /
+      my-cluster-kafka-bootstrap.kafka:9093 --topic my-topic --from-beginning \
+      --consumer-property 'security.protocol=SSL'
+
+We should get a disconnected error. In the Kafka broker log we should see 'Failed authentication' error message.
+
+Let's fix that with proper authentication configuration:
+
+    export OAUTH_CLIENT_ID=kafka-consumer
+    export OAUTH_CLIENT_SECRET=<SECRET_FOR_KAFKA_CONSUMER_FROM_KEYCLOAK_CONSOLE>
+    export PASSWORD=truststorepassword
+    export KAFKA_OPTS=" \
+      -Djavax.net.ssl.trustStore=/opt/kafka/certificates/kafka-client-truststore.p12 \
+      -Djavax.net.ssl.trustStorePassword=$PASSWORD \
+      -Djavax.net.ssl.trustStoreType=PKCS12"
+      
+    bin/kafka-console-consumer.sh --bootstrap-server \
+      my-cluster-kafka-bootstrap.kafka:9093 --topic my-topic --from-beginning \
+      --consumer-property 'security.protocol=SASL_SSL' \
+      --consumer-property 'sasl.mechanism=OAUTHBEARER' \
+      --consumer-property 'sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required;' \
+      --consumer-property 'sasl.login.callback.handler.class=io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler' 
+
+Now, all should work as expected.
+Our client should receive all the messages in the topic, and if we keep it running would receive any additional produced messages.
+
+
+
+## Conclusion
+
+OAuth 2.0 token-based authentication allows you to use central management of users and security policies by way of an external security server, and is relatively easy to setup.
+
+What we have implemented thus far is initial support for authentication. 
+For next steps, we are looking at roles-based authorization, which could be used instead of the default ACL based authorization to centrally manage not only user identities, but also the permissions they have on Kafka resources.
+
+OAuth2 support is part of Strimzi 0.14, and you can use it with Strimzi brokers, Kafka Mirror Maker, Kafka Connect, Kafka Bridge, and with your custom Kafka client applications.
+Give it a try, and let us know what you think.
```

Comment:
I would add what we have in other blog posts:

"If you like Strimzi, star us on [GitHub](https://github.com/strimzi/strimzi-kafka-operator) and follow us on [Twitter](https://twitter.com/strimziio) to make sure you don't miss any of our future blog posts!"